fairseq/models/wav2vec/wav2vec2.py

```diff
@@ -454,21 +473,23 @@ def sample_negatives(self, y, num):
         y = y.view(-1, fsz)  # BTC => (BxT)C
 
         cross_high = tsz * bsz
-        high = tsz
+        high = num
+        # FIXME: there's a problem here w/ tsz and num
+        # this assumes y is shrunk at this point.
         with torch.no_grad():
             assert high > 1, f"{bsz,tsz,fsz}"
 
             if self.n_negatives > 0:
                 tszs = (
-                    buffered_arange(num)
+                    buffered_arange(tsz)
                     .unsqueeze(-1)
                     .expand(-1, self.n_negatives)
                     .flatten()
                 )
 
-                neg_idxs = torch.randint(
-                    low=0, high=high - 1, size=(bsz, self.n_negatives * num)
-                )
+                neg_idxs = torch.randint(low=0, high=high-1, size=(bsz, self.n_negatives * tsz))
```

Comment:
here, for xla, you should loop over bsz and sample negatives individually for each example setting high to be tsz-sum(padding[b]) as we discussed. otherwise you might be sampling negatives from states that are padded. i guess padding[b] would come from padding_counts which is currently not used